ParserException when bulk-requesting resources by ID with crnk-operations

Hello,

With Crnk v3.1.20190810073916 I was able to use the operations endpoint to do bulk GET requests for single resources by ID, but with v3.2.20191227143629 this fails when I use the "include" param.

Request body:
```
[{
  "op": "GET",
  "path":"metadata/729da2da-4cb0-4162-811a-219ab239ed03?include=acMetadataCreator,managedAttributeMap",
  "value":{}
}]
```

Response:
```
{
    "errors": [
        {
            "status": "500",
            "title": "INTERNAL_SERVER_ERROR",
            "detail": "not a valid uuid"
        }
    ]
}
```

This only works if I omit the "include".

Operations seems to parse the "path" differently than if I request that path directly in the URL, which still works.

Log output:
```
io.crnk.core.engine.parser.ParserException: not a valid uuid
        at io.crnk.core.engine.parser.DefaultStringParsers$11.parse(DefaultStringParsers.java:101) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.parser.DefaultStringParsers$11.parse(DefaultStringParsers.java:95) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.parser.TypeParser.parse(TypeParser.java:136) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.information.resource.ResourceInformation$1.parse(ResourceInformation.java:113) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.information.resource.ResourceInformation.parseIdString(ResourceInformation.java:574) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.internal.dispatcher.path.PathBuilder.parseIds(PathBuilder.java:51) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.internal.dispatcher.path.PathBuilder.parseIdPath(PathBuilder.java:123) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.internal.dispatcher.path.PathBuilder.parseResourcePath(PathBuilder.java:118) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.internal.dispatcher.path.PathBuilder.build(PathBuilder.java:101) ~[crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.OperationsModule.executeOperations(OperationsModule.java:206) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.OperationsModule$DefaultOperationFilterChain.doFilter(OperationsModule.java:448) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.TransactionOperationFilter.lambda$filter$0(TransactionOperationFilter.java:19) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.spring.jpa.SpringTransactionRunner$1.doInTransaction(SpringTransactionRunner.java:30) ~[crnk-setup-spring-3.2.20191227143629.jar:na]
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140) ~[spring-tx-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at io.crnk.spring.jpa.SpringTransactionRunner.doInTransaction(SpringTransactionRunner.java:25) ~[crnk-setup-spring-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.TransactionOperationFilter.filter(TransactionOperationFilter.java:19) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.OperationsModule$DefaultOperationFilterChain.doFilter(OperationsModule.java:452) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.OperationsModule.apply(OperationsModule.java:116) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.operations.server.OperationsRequestProcessor.process(OperationsRequestProcessor.java:44) ~[crnk-operations-3.2.20191227143629.jar:na]
        at io.crnk.core.engine.internal.http.HttpRequestDispatcherImpl.process(HttpRequestDispatcherImpl.java:88) [crnk-core-3.2.20191227143629.jar:na]
        at io.crnk.servlet.CrnkFilter.doFilter(CrnkFilter.java:85) [crnk-setup-servlet-3.2.20191227143629.jar:na]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119) [spring-web-5.2.1.RELEASE.jar:5.2.1.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:861) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1579) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_144]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_144]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.27.jar:9.0.27]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_144]
Caused by: java.lang.NumberFormatException: For input string: "219ab239ed03?include=acMetadataCreator"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65) ~[na:1.8.0_144]
        at java.lang.Long.parseLong(Long.java:589) ~[na:1.8.0_144]
        at java.lang.Long.valueOf(Long.java:776) ~[na:1.8.0_144]
        at java.lang.Long.decode(Long.java:928) ~[na:1.8.0_144]
        at java.util.UUID.fromString(UUID.java:206) ~[na:1.8.0_144]
        at io.crnk.core.engine.parser.DefaultStringParsers$11.parse(DefaultStringParsers.java:99) ~[crnk-core-3.2.20191227143629.jar:na]
        ... 50 common frames omitted
```

--- a/crnk-documentation/src/docs/asciidoc/operations.adoc
+++ b/crnk-documentation/src/docs/asciidoc/operations.adoc
@@ -169,10 +169,9 @@ to have `application/json-patch+json`, otherwise the `OperationsModule` will ign
 
 The current limitations of the implementation are:
 
-- So far does not support bulk `GET` operations.
 - Does so far not support bulk update of relationships.
 
-With support for `POST`, `PATCH` and `DELETE` operations the most important building blocks should be in place.
+With support for `GET` (find-one only), `POST`, `PATCH` and `DELETE` operations the most important building blocks should be in place.
 The limitations are expected to be addressed at some point as well, contributions welcomed.
 
 There is experimental support for <<resource_bulk_repository,BulkResourceRepository>>. If multiple changes to the

--- a/crnk-operations/src/main/java/io/crnk/operations/client/OperationsCall.java
+++ b/crnk-operations/src/main/java/io/crnk/operations/client/OperationsCall.java
@@ -57,6 +57,16 @@ public class OperationsCall {
 		queuedOperations.add(queuedOperation);
 	}
 
+	public void add(HttpMethod method, String path) {
+		Operation operation = new Operation();
+		operation.setOp(method.toString());
+		operation.setPath(path);
+
+		QueuedOperation queuedOperation = new QueuedOperation();
+		queuedOperation.operation = operation;
+		queuedOperations.add(queuedOperation);
+	}
+
 	protected String computePath(HttpMethod method, Resource resource) {
 		if (method == HttpMethod.POST) {
 			return resource.getType();

--- a/crnk-operations/src/main/java/io/crnk/operations/server/OperationsModule.java
+++ b/crnk-operations/src/main/java/io/crnk/operations/server/OperationsModule.java
@@ -151,7 +151,7 @@ public class OperationsModule implements Module {
     private void enrichTypeIdInformation(List<Operation> operations, QueryContext queryContext) {
         ResourceRegistry resourceRegistry = moduleContext.getResourceRegistry();
         for (Operation operation : operations) {
-            if (operation.getOp().equalsIgnoreCase(HttpMethod.DELETE.toString())) {
+            if (Arrays.asList(HttpMethod.DELETE.toString(), HttpMethod.GET.toString()).contains(operation.getOp())) {
                 String path = OperationParameterUtils.parsePath(operation.getPath());
                 PathBuilder pathBuilder = new PathBuilder(resourceRegistry, moduleContext.getTypeParser());
                 JsonPath jsonPath = pathBuilder.build(path, queryContext);
@@ -203,12 +203,13 @@ public class OperationsModule implements Module {
 
             Operation operation = orderedOperation.getOperation();
             String method = operation.getOp();
-            JsonPath path = pathBuilder.build(operation.getPath(), queryContext);
-            orderedOperation.setPath(path);
-            if (path == null) {
+            String path = OperationParameterUtils.parsePath(operation.getPath());
+            JsonPath jsonPath = pathBuilder.build(path, queryContext);
+            orderedOperation.setPath(jsonPath);
+            if (jsonPath == null) {
                 throw new BadRequestException("invalid path: " + operation.getPath());
             }
-            String type = path.getRootEntry().getResourceInformation().getResourceType();
+            String type = jsonPath.getRootEntry().getResourceInformation().getResourceType();
 
             if (!method.equals(bulkMethod) || !type.equals(bulkType)) {
                 if (!bulk.isEmpty()) {

--- a/crnk-operations/src/test/java/io/crnk/operations/OperationsDeleteTest.java
+++ b/crnk-operations/src/test/java/io/crnk/operations/OperationsDeleteTest.java
@@ -88,6 +88,33 @@ public class OperationsDeleteTest extends AbstractOperationsTest {
   }
 
   @Test
+  public void checkDeleteByPath() {
+    ResourceRepository<PersonEntity, UUID> personRepo = client.getRepositoryForType(PersonEntity.class);
+
+    PersonEntity person1 = newPerson("1");
+    PersonEntity person2 = newPerson("2");
+
+    OperationsClient operationsClient = new OperationsClient(client);
+    OperationsCall call = operationsClient.createCall();
+    call.add(HttpMethod.POST, person1);
+    call.add(HttpMethod.POST, person2);
+    call.execute();
+
+    QuerySpec querySpecBeforeDelete = new QuerySpec(PersonEntity.class);
+    ResourceList<PersonEntity> personsBeforeDelete = personRepo.findAll(querySpecBeforeDelete);
+    Assert.assertEquals(2, personsBeforeDelete.size());
+
+    call = operationsClient.createCall();
+    call.add(HttpMethod.DELETE, "person/" + person1.getId());
+    call.add(HttpMethod.DELETE, "person/" + person2.getId());
+    call.execute();
+
+    QuerySpec querySpecAfterDelete = new QuerySpec(PersonEntity.class);
+    ResourceList<PersonEntity> personsAfterDelete = personRepo.findAll(querySpecAfterDelete);
+    Assert.assertEquals(0, personsAfterDelete.size());
+  }
+
+  @Test
   public void checkExperimentalBulkDelete() {
     Task task1 = newTask("1");
     Task task2 = newTask("2");

--- a/crnk-operations/src/test/java/io/crnk/operations/OperationsGetTest.java
+++ b/crnk-operations/src/test/java/io/crnk/operations/OperationsGetTest.java
@@ -1,0 +1,80 @@
+package io.crnk.operations;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.UUID;
+
+import javax.persistence.EntityManager;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import io.crnk.core.engine.http.HttpMethod;
+import io.crnk.core.repository.ResourceRepository;
+import io.crnk.operations.client.OperationsCall;
+import io.crnk.operations.client.OperationsClient;
+import io.crnk.operations.model.MovieEntity;
+import io.crnk.operations.model.PersonEntity;
+import io.crnk.spring.jpa.SpringTransactionRunner;
+
+public class OperationsGetTest extends AbstractOperationsTest {
+
+  protected ResourceRepository<MovieEntity, UUID> movieRepo;
+
+  PersonEntity person1;
+  PersonEntity person2;
+
+  @Before
+  @Override
+  public void setUp() throws Exception {
+    super.setUp();
+    movieRepo = client.getRepositoryForType(MovieEntity.class);
+
+    SpringTransactionRunner transactionRunner = context.getBean(SpringTransactionRunner.class);
+    transactionRunner.doInTransaction(() -> {
+      EntityManager em = context.getBean(EntityManagerProducer.class).getEntityManager();
+
+      person1 = newPerson("P1");
+      em.persist(person1);
+
+      person2 = newPerson("P2");
+      em.persist(person2);
+
+      MovieEntity movie = newMovie("test");
+      em.persist(movie);
+
+      movie.setDirectors(new HashSet<>(Arrays.asList(person1, person2)));
+
+      return null;
+    });
+  }
+
+  @Test
+  public void checkGet() {
+    OperationsClient operationsClient = new OperationsClient(client);
+    OperationsCall call = operationsClient.createCall();
+    call.add(HttpMethod.GET, "person/" + person1.getId());
+    call.add(HttpMethod.GET, "person/" + person2.getId());
+    call.execute();
+
+    Assert.assertEquals("P1", call.getResponse(0).getSingleData().get().getAttributes().get("name").asText());
+    Assert.assertEquals("P2", call.getResponse(1).getSingleData().get().getAttributes().get("name").asText());
+  }
+
+  @Test
+  public void checkGetWithIncludeParam() {
+    OperationsClient operationsClient = new OperationsClient(client);
+    OperationsCall call = operationsClient.createCall();
+    call.add(HttpMethod.GET, "person/" + person1.getId() + "?include=directedMovies");
+    call.add(HttpMethod.GET, "person/" + person2.getId() + "?include=directedMovies");
+    call.execute();
+
+    // directedMovies should have been included:
+    Assert.assertEquals(1, call.getResponse(0).getSingleData().get().getRelationships().get("directedMovies")
+        .getCollectionData().get().size());
+    Assert.assertEquals(1, call.getResponse(1).getSingleData().get().getRelationships().get("directedMovies")
+        .getCollectionData().get().size());
+  }
+
+}

